### DISCUSSION TITLE
How to get UserHandle by CID/URI ?

### DISCUSSION THREAD

**Discussion by CommanderWaterford** [bluesky-social/atproto#1883] [How to do it?] [answered]
Hi, via the firehose API I got a JSON Object which looks like the one below for example - how can I get the UserHandle of this Skeet via atproto? Any help greatly appreciated.

`Record Object: {
  "text": "Salut copain du nuldi 😆",
  "$type": "app.bsky.feed.post",
  "langs": [
    "fr"
  ],
  "reply": {
    "root": {
      "cid": "bafyreibfd77vb2setujncomtz3j6xswrmiuxlykora6nogxbr4arhqu2ye",
      "uri": "at://did:plc:dbedpmb6db7fsbqcasqxlcum/app.bsky.feed.post/3keqzgxzats25"
    },
    "parent": {
      "cid": "bafyreibfd77vb2setujncomtz3j6xswrmiuxlykora6nogxbr4arhqu2ye",
      "uri": "at://did:plc:dbedpmb6db7fsbqcasqxlcum/app.bsky.feed.post/3keqzgxzats25"
    }
  },
  "createdAt": "2023-11-22T09:09:52.420Z"
}
`

**Comment by bnewbold** [comment, marked as answer]
You can't find the identity which created this post from just the record, you need the other metadata from the `#commit` message on the event stream. Specifically, the `repo` field of the `com.atproto.sync.subscribeRepos#commit` message is the DID of the repo (DID of the account).

Starting from a DID, you can resolve the identity metadata by resolving the DID document (either did:plc or did:web), extracting any handle for the account, and resolving the handle to confirm it is valid (points back to the correct DID). Usually a service would use a library to do this identity resolution, and also cache things locally for up to 24 hours to reduce the number of network requests.

**Comment by CommanderWaterford** [comment]
1st and foremost Thank you for your answer although I admit that I did understand perhaps 50% ;-). I am wondering how Firesky does realize this for example. Caching is not an option. So bluntly atproto does not offer any solution via API to get it (which is not rate limited in the end).

**Reply by bnewbold** [reply]
There is an API endpoint `com.atproto.repo.describeRepo` which takes a DID or handle and returns the DID doc, handle, and a flag whether the handle is valid.
